### ISSUE TITLE
À trous 1D convolution slow in certain scenarios 

### ISSUE THREAD

**Issue by zbitouzakaria** [tensorflow/tensorflow#21923]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: 3.5.2
- **CUDA/cuDNN version**: 9.0.176
- **GPU model and memory**: CPU
- **Exact command to reproduce**: See below
- **Bazel version**: N/A
- **Mobile device**: N/A

### Describe the problem
Running

``
tf.nn.atrous_conv2d(images, filters, rate=dilation_rate, padding="SAME")
``

and

``
tf.nn.convolution(images, filters, dilation_rate=[dilation_rate, dilation_rate], padding="SAME")
``

takes much more time to run than:

``
tf.nn.convolution(images, filters, dilation_rate=[1, dilation_rate], padding="SAME")
``

for `1D` convolutions.

I don't think this is supposed to behave like this, provided that the input data is one dimensional along the width or the height (i.e. ``shape = [batch, 1, width, channels] or [batch, height, 1, channels]``) since in this case, no dilation is needed on the corresponding axis.

A person using `tf.nn.atrous_conv2d` for `1D` "à trous" convolutions might never notice the issue here, and his network will be really slow to train.

### Source code / logs
A small snippet to compare the performance of the different methods

```python
import tensorflow as tf
import numpy as np
import time

# Scenario
batch_size = 100
channels = 3
o_channels = 10
image_size = 160
filter_size = 3

# Dummy images and filters
images = tf.random_normal(shape=[batch_size, 1, image_size, channels],  dtype=tf.float32)
filters = tf.random_normal(shape=[1, filter_size, channels, o_channels], dtype=tf.float32)

# Ops definition
normal_conv = tf.nn.convolution(images, filters, padding="SAME")
atrous_convs = []
atrous_convs_wrapper = []
for i in [5, 10, 20, 50]:
    atrous_convs.append({"dilation": i, "op": tf.nn.convolution(images, filters, dilation_rate=[1, i], padding="SAME")})
    atrous_convs_wrapper.append({"dilation": i, "op": tf.nn.convolution(images, filters, dilation_rate=[i, i], padding="SAME")})
    # Or
    # atrous_convs_wrapper.append({"dilation": i, "op": tf.nn.atrous_conv2d(images, filters, rate=i, padding="SAME")})
# Test
with tf.Session() as sess:
    repeats = 50
    
    for conv in atrous_convs:
        op1 = conv["op"]
        op2 = [elt for elt in atrous_convs_wrapper if elt["dilation"] == conv["dilation"]][0]["op"]
        op1_res, op2_res = sess.run([op1, op2])
        np.testing.assert_equal(op1_res, op2_res)
    
    # Benchmark normal method
    start = time.time()
    for _ in range(repeats):
        _ = sess.run(normal_conv)
    end = time.time()
    normal_conv_time = int((end - start) / repeats * 1000)

    # Benchmark atrous conv method
    for conv in atrous_convs:
        start = time.time()
        for _ in range(repeats):
            _ = sess.run(conv["op"])
        end = time.time()
        conv.update({"time": int((end - start) / repeats * 1000)})
    
    for conv in atrous_convs_wrapper:
        start = time.time()
        for _ in range(repeats):
            _ = sess.run(conv["op"])
        end = time.time()
        conv.update({"time": int((end - start) / repeats * 1000)})

    print("Normal conv: {}ms".format(normal_conv_time))
    for conv in atrous_convs:
        print("Atrous conv w/ dilation {}: {}ms".format(conv["dilation"], conv["time"]))
    for conv in atrous_convs_wrapper:
        print("Atrous conv(wrapper) w/ dilation {}: {}ms".format(conv["dilation"], conv["time"]))

```

returns

```
Normal conv: 1ms
Atrous conv w/ dilation 5: 1ms
Atrous conv w/ dilation 10: 2ms
Atrous conv w/ dilation 20: 1ms
Atrous conv w/ dilation 50: 2ms
Atrous conv(wrapper) w/ dilation 5: 4ms
Atrous conv(wrapper) w/ dilation 10: 9ms
Atrous conv(wrapper) w/ dilation 20: 15ms
Atrous conv(wrapper) w/ dilation 50: 47ms
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
Mobile device

**Comment by zbitouzakaria** [comment]
@tensorflowbutler Done :)

**Comment by zbitouzakaria** [comment]
The best might be to simply specify it in the documentation. Something along theses lines for the [atrous_conv2d doc](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d):

- Before

> This function is a simpler wrapper around the more general tf.nn.convolution, and exists only for backwards compatibility. You can use tf.nn.convolution to perform 1-D, 2-D, or 3-D atrous convolution.


- After

> This function is a simpler wrapper around the more general tf.nn.convolution, and exists only for backwards compatibility. Make sure to use tf.nn.convolution to perform 1-D, 2-D, or 3-D atrous convolutions, by specifying the proper dilation rate for your input.

> Be especially wary not to use this function to perform 1-D atrous convolutions, since we zero-pad the input tensor to have a size multiple of the dilation rate along the corresponding dimensions, which will affect the performance of the operation.

Or you could put up a warning in the atrous_conv2d function if you notice that the input tensor is one dimensional along the width or height axis.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by harshini-gadige** [comment]
@skye Hi, is there any update made in the documentation as per above ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gadagashwini** [comment]
@zbitouzakaria,

> If the rate parameter is equal to one, it performs regular 2-D convolution. If the rate parameter is greater than one, **it performs convolution with holes, sampling the input values every rate pixels in the height and width dimensions.** This is equivalent to convolving the input with a set of upsampled filters, produced by inserting rate - 1 zeros between two consecutive values of the filters along the height and width dimensions, hence the name atrous convolution or convolution with holes. 
 Does this explanation not suffice, given [here](https://www.tensorflow.org/api_docs/python/tf/nn/atrous_conv2d). Thank you!
